infra/relayer/src/purse.js

```diff
@@ -0,0 +1,682 @@
+ * Ethereum and BIP44 discussion - https://github.com/ethereum/EIPs/issues/84
+ *
+ * TODO
+ * ----
+ * - Correct redis stored nonce if a tx fails?
+ * - Locks and support for parallelism so multiple instances can work at the same time?
+ * - Auto-scale children when all accounts hit the max pending setting?
+ */
+const { promisify } = require('util')
+const redis = require('redis')
+const BN = require('bn.js')
+const bip39 = require('bip39')
+const hdkey = require('ethereumjs-wallet/hdkey')
+const {
+  stringToBN,
+  numberToBN,
+  getBIP44Path,
+  bufferToHex,
+  sendRawTransaction
+} = require('./util')
+const logger = require('./logger')
+
+const REDIS_RETRY_TIMEOUT = 30000
+const REDIS_RETRY_DELAY = 500
+const DEFAULT_CHILDREN = 5
+const DEFAULT_MAX_PENDING_PER_ACCOUNT = 3
+const ZERO = new BN('0', 10)
+const BASE_FUND_VALUE = new BN('50000000000000000', 10) // 0.05 Ether
+const MIN_CHILD_BALANCE = new BN('1000000000000000', 10) // 0.001 Ether
+const MAX_GAS_PRICE = new BN('20000000000', 10) // 20 gwei
+const REDIS_TX_COUNT_PREFIX = 'txcount_'
+
+async function tick(wait = 1000) {
+  return new Promise(resolve => setTimeout(() => resolve(true), wait))
+}
+
+class Account {
+  constructor({ txCount, pendingCount = 0, wallet, funded }) {
+    this.txCount = txCount
+    this.pendingCount = pendingCount // needed?
+    this.wallet = wallet
+    this.funded = funded
+  }
+}
+
+/**
+ * Purse is an account abstraction that allows relayer to feed it arbitrary transactions to send
+ * with whatever account it has avilable.  It does not block and wait for transactions to be mined.
+ * You can not choose which account to send from.  No transactions, except those to fund the
+ * derived children will be sent from the master account (of the mnemonic provided).
+ *
+ * Usage
+ * -----
+ * a = Accounts({ web3, mnemonic: 'one two three' })
+ * await a.init()
+ * await a.sendTx({ from: you, to: me, data: '0xdeadbeef' })
+ */
+class Purse {
+  constructor({
+    web3,
+    mnemonic,
+    children = DEFAULT_CHILDREN,
+    autofundChildren = false,
+    redisHost = 'redis://localhost:6379/0',
+    maxPendingPerAccount = DEFAULT_MAX_PENDING_PER_ACCOUNT
+  }) {
+    if (!web3 || !mnemonic) {
+      throw new Error('missing required parameters')
+    }
+
+    this.web3 = web3
+    this.mnemonic = mnemonic
+    this._childrenToCreate = children
+    this.autofundChildren = autofundChildren
+    this.maxPendingPerAccount = maxPendingPerAccount
+
+    this.ready = false
+    this.masterKey = null
+    this.masterWallet = null
+
+    this.gasPrice = new BN('3000000000', 10) // 3 gwei TODO: floating source?
+
+    this.children = []
+    this.accounts = {}
+    // Keep track of transactions that are waiting to be mined
+    this.pendingTransactions = {}
+    // Counter for tx rebroadcasts
+    this.rebroadcastCounters = {}
+    // Callbacks for when receipts show up
+    this.receiptCallbacks = {}
+
+    this.rclient = this._setupRedis(redisHost)
```

Comment:
Would it be a lot of work to have redis as optional so it's still possible to run the relayer locally without any other dependencies?

Reply:
(just reading through... maybe it's already optional)

Reply:
Yeah, it's optional.  For example, the marketplace tests aren't configured use Redis.